icecreamshop/icecreamshop/src/test/java/com/summerschool/icecreamshop/product/controller/ProductControllerTest.java

```diff
@@ -22,34 +29,131 @@
+
+        Product product = buildValidProduct();
+        Integer idProduct = product.getId();
+
+        Mockito.when(productService.findById(idProduct)).thenReturn(Optional.of(product));
+        Mockito.when(productService.update(product)).thenReturn(product);
+
+        MockHttpServletRequestBuilder builder = buildMockPatch(idProduct, product);
+
+        this.mockMvc.perform(builder)
+                .andExpect(MockMvcResultMatchers.status()
+                        .isOk());
+    }
+
+    @Test
+    void givenInvalidProduct_WhenUpdate_ThenStatusIsBadRequest() throws Exception {
+
+        Product product = buildValidProduct();
+        product.setTitle(null);
+        Integer idProduct = product.getId();
+
+        Mockito.when(productService.findById(idProduct)).thenReturn(Optional.of(product));
+        Mockito.when(productService.update(product)).thenReturn(product);
+
+        MockHttpServletRequestBuilder builder = buildMockPatch(idProduct, product);
+
+        this.mockMvc.perform(builder)
+                .andExpect(MockMvcResultMatchers.status()
+                        .isBadRequest());
+    }
+
+
+    @Test
+    void givenNotExistingProduct_WhenUpdate_ThenStatusIsNotFound() throws Exception {
+
+        Product product = buildValidProduct();
+        Integer idProduct = product.getId();
+
+        Mockito.when(productService.findById(idProduct)).thenReturn(Optional.empty());
+        Mockito.when(productService.update(product)).thenReturn(product);
+
+        MockHttpServletRequestBuilder builder = buildMockPatch(idProduct, product);
+
+        this.mockMvc.perform(builder)
+                .andExpect(MockMvcResultMatchers.status()
+                        .isNotFound());
+    }
+
+    @Test
+    void givenIdFromPathIsNotEqualWithIdFromProduct_WhenUpdate_ThenStatusIsBadRequest() throws Exception {
+        // given id from path != id from product when update
+
+        Product product = buildValidProduct();
+        Integer idProduct = product.getId() - 1;
+
+        Mockito.when(productService.findById(idProduct)).thenReturn(Optional.empty());
+        Mockito.when(productService.update(product)).thenReturn(product);
+
+        MockHttpServletRequestBuilder builder = buildMockPatch(idProduct, product);
+
+        this.mockMvc.perform(builder)
+                .andExpect(MockMvcResultMatchers.status()
+                        .isBadRequest());
+    }
+
+    private Product buildValidProduct() {
+        Product product = new Product();
+        product.setId(1);
+        product.setTitle("Vanilla ice cream");
+        product.setShortDescription("Tasty");
+        product.setIngredients("Milk, vanilla extract");
+        product.setQuantity(100);
+        product.setAlergens("Milk");
+        product.setPrice(5.99f);
+        product.setCurrency("lei");
+        product.setType(Product.Type.ICE_CREAM);
+        product.setCategoryId(100);
+        return product;
+    }
```

Comment:
Instead of calling buildValidProduct() in each test method you can make use of a setUp() or init() method in which you can initialise the objects that you need for the tests.

https://www.baeldung.com/junit-5 (Chapter 4.1)

Reply:
Do I create this methd in Product class?